inc/elasticsearch_packages/namespace.php

```diff
@@ -0,0 +1,432 @@
+}
+
+/**
+ * Get the list of packages to remove.
+ *
+ * @return array
+ */
+function get_packages_to_remove() : array {
+	return get_site_option( 'altis_search_packages_to_remove', [] );
+}
+
+/**
+ * Check and update the package association status.
+ *
+ * @param string $package_id The package ID to look up.
+ * @param string $slug The package slug.
+ * @param boolean $for_network Whether this is a network level package.
+ * @param string|null $existing_package_id An existing package ID to be removed.
+ * @return void
+ */
+function on_check_package_status( string $package_id, string $slug, bool $for_network = false, ?string $existing_package_id = null ) {
+
+	$client = get_es_client();
+
+	$real_package_id = str_replace( 'analyzers/', '', $package_id );
+
+	// Set scheduled event hook arguments.
+	$check_status_hook_args = [ $package_id, $slug, $for_network, $existing_package_id ];
+	$update_index_hook_args = [ $for_network, strpos( $slug, 'user-dictionary' ) !== false ];
+
+	try {
+		$packages = $client->listDomainsForPackage( [
+			'PackageID' => $real_package_id,
+		] );
+
+		// Package may not have finished copying over yet.
+		if ( empty( $packages['DomainPackageDetailsList'] ) ) {
+			$packages = $client->describePackages( [
+				'Filters' => [
+					[
+						'Name' => 'PackageID',
+						'Value' => [ $real_package_id ],
+					],
+				],
+			] );
+
+			// We shouldn't get to this point without having run the create package step
+			// but this ensures we handle that scenario in case it comes up.
+			if ( empty( $packages['PackageDetailsList'] ) ) {
+				throw new Exception( sprintf( 'Elasticsearch package %s not found. Try saving it again or contact support.', $package_id ) );
+			}
+
+			$package = $packages['PackageDetailsList'][0];
+			$status = $package['PackageStatus'];
+
+			// Associate the package with the ES domain.
+			// NOTE: This process is async so cannot be applied immediately.
+			if ( $package['PackageStatus'] === 'AVAILABLE' ) {
+				$result = $client->associatePackage( [
+					'DomainName' => get_elasticsearch_domain(), // required.
+					'PackageID' => $real_package_id, // required.
+				] );
+				$package = $result['DomainPackageDetails'];
+				$status = $package['DomainPackageStatus'];
+			}
+		} else {
+			// Package is currently being associated or is active.
+			$package = $packages['DomainPackageDetailsList'][0];
+			$status = $package['DomainPackageStatus'];
+		}
+
+		// Update the stored status.
+		if ( $for_network ) {
+			update_site_option( "altis_search_package_status_{$slug}", $status );
+			update_site_option( "altis_search_package_error_{$slug}", $package['ErrorDetails']['ErrorMessage'] ?? null );
+		} else {
+			update_option( "altis_search_package_status_{$slug}", $status );
+			update_option( "altis_search_package_error_{$slug}", $package['ErrorDetails']['ErrorMessage'] ?? null );
+		}
+
+		// If the package is now active then update the index settings.
+		if ( $status === 'ACTIVE' ) {
```

Comment:
Ideally these conditions should be switched so that the "happy path" is the main flow through the code (and so that the conditional itself is shorter)